### PR TITLE
ARROW-1722: [C++] Add linting script to find C++/CLI incompatibilities

### PR DESCRIPTION
**Pull request by wesm**
This also fixes some issues that the script found

### CODE REVIEW

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-2810 -- I was surprised to find Plasma installs its generated Flatbuffers files

File: cpp/build-support/lint_cpp_cli.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+for dirpath, _, filenames in os.walk(arguments.source_path):
+    for filename in filenames:
+        full_path = os.path.join(dirpath, filename)
+
+        exclude = False
+        for exclusion in EXCLUSIONS:
+            if exclusion in full_path:
+                # print('Skipping excluded header: {0}'.format(filename))
```
Comment (xhochy) [inline cpp/build-support/lint_cpp_cli.py]: Remove or replace with `logger.debug`

**Review by xhochy — COMMENTED** [review]
LGTM except a small comment

File: cpp/build-support/lint_cpp_cli.py (hunk 1 above)
Comment (wesm) [reply cpp/build-support/lint_cpp_cli.py]: Fixed

**Review by xhochy — APPROVED** [review]

File: cpp/build-support/lint_cpp_cli.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import os
+
+parser = argparse.ArgumentParser(
+    description="Check for illegal headers for C++/CLI applications")
+parser.add_argument("source_path",
+                    help="Path to source code")
+arguments = parser.parse_args()
+
+
+_STRIP_COMMENT_REGEX = re.compile('(.+)?(?=//)')
```
Comment (kou) [inline cpp/build-support/lint_cpp_cli.py]: `'(.+?)?(?=//)'` will be better for `a // b // c` case but it's too trivial. We will never write such code :-)

**Review by kou — APPROVED** [review]
+1
I'll merge this.

**Pull request closed without merging**